packages/nextjs/src/server/types.ts

```diff
@@ -0,0 +1,40 @@
+import { AuthErrorReason, ClerkBackendAPI, Organization, Session, User } from '@clerk/backend-core';
+import { ClerkJWTClaims, ServerGetToken } from '@clerk/types';
+import { IncomingMessage } from 'http';
+import { NextApiRequest } from 'next';
+import { NextApiRequestCookies } from 'next/dist/server/api-utils';
+import { NextRequest } from 'next/server';
+
+// Request contained in GetServerSidePropsContext, has cookies but not query
+type GsspRequest = IncomingMessage & {
+  cookies: NextApiRequestCookies;
+};
+
+export type RequestLike = NextRequest | NextApiRequest | GsspRequest;
+
+export const SESSION_COOKIE_NAME = '__session';
+export const AUTH_RESULT = 'Auth-Result';
+export const NEXT_REWRITE_HEADER = 'x-middleware-rewrite';
+export const NEXT_RESUME_HEADER = 'x-middleware-next';
+export const NEXT_REDIRECT_HEADER = 'Location';
+
+// TODO consolidate AuthData declarations across repo
+export type AuthData = {
+  sessionId: string | null;
+  session: Session | undefined | null;
+  userId: string | null;
+  user: User | undefined | null;
```

Comment:
❓ Since we don't load the actual resources in the context of the middleware, which is the scenario that we get a `session`, a `user`, or an `organization` object in the `AuthData`?

Reply:
We don't but it's confusing to have so many different AuthData declarations so I'm using the superset here.

Otherwise we'd have to generate different signed out data for this case as well.

I don't think the complexity is worth it.

Reply:
I am skeptical whether this will create expectations to the end users about getting the full objects. Maybe we can just Pick the rest.

Reply:
It would be nice if we had AR-like `association.loaded?` semantics.

Reply:
Isn't that a bit over-engineering given that we have the undefined type?

Reply:
Maybe if we had the following notion:
- `null` = known not to exist
- `undefined` = not populated, you'll need to load it

Reply:
Anyway, narrowed down to only the fields we are setting, for now.